### PR TITLE
Add VectorizedBaseImageAugmentation layer

### PR DESCRIPTION
**Pull request by LukeWood**
Figure_1

Old non-graphmode basescale is like 100s runtime so not even comparable.

### CODE REVIEW

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+        Returns:
+          output 3D tensor, which will be forward to `layer.call()`.
+        """
+        raise NotImplementedError()
+
+    def augment_keypoints(self, keypoints, transformations, **kwargs):
+        """Augment a batch of keypoints for one image during training.
+
+        Args:
+          keypoints: 3D keypoints input tensor to the layer. Forwarded from
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:188]: shape?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        """
+        return self.augment_labels(targets, transformations)
+
+    def augment_bounding_boxes(self, bounding_boxes, transformations, **kwargs):
+        """Augment bounding boxes for one image during training.
+
+        Args:
+          image: 3D image input tensor to the layer. Forwarded from
+            `layer.call()`.
+          bounding_boxes: 2D bounding boxes to the layer. Forwarded from
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:173]: shape?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        Returns:
+          output 2D tensor, which will be forward to `layer.call()`.
+        """
+        return self.augment_labels(targets, transformations)
+
+    def augment_bounding_boxes(self, bounding_boxes, transformations, **kwargs):
+        """Augment bounding boxes for one image during training.
+
+        Args:
+          image: 3D image input tensor to the layer. Forwarded from
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:171]: 4D -- right?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 2 above)
Comment (ianstenbit) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:173]: Also isn't this 3D?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+          transformations: The transformations object produced by
+            `get_random_transformations`. Used to coordinate the randomness
+            between image, label, bounding box, keypoints, and segmentation mask.
+
+        Returns:
+          output 2D tensor, which will be forward to `layer.call()`.
+        """
+        raise NotImplementedError()
+
+    def augment_targets(self, targets, transformations, **kwargs):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:153]: Shouldn't we just drop this, since it's just wrapping around `augment_labels`?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        Returns:
+          output 4D tensor, which will be forward to `layer.call()`.
+        """
+        raise NotImplementedError()
+
+    def augment_labels(self, labels, transformations, **kwargs):
+        """Augment a batch of  labels during training.
+
+        Args:
+          label: 2D label to the layer. Forwarded from `layer.call()`.
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:143]: shape?

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+    def augment_keypoints(self, keypoints, transformations, **kwargs):
+        return keypoints + transformations[:, None, None]
+
+    def augment_segmentation_masks(self, segmentation_masks, transformations, **kwargs):
+        return segmentation_masks + transformations[:, None, None, None]
+
+
+class BatchedBaseImageAugmentationLayerTest(tf.test.TestCase):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py:55]: Vectorized

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py:1]: 2023 -- also, should this be marked as branched from the old base layer test for better blame history?

File: keras_cv/layers/preprocessing/grayscale.py [hunk 8]
```diff
@@ -14,13 +14,13 @@
...
-from keras_cv.layers.preprocessing.base_image_augmentation_layer import (
-    BaseImageAugmentationLayer,
+from keras_cv.layers.preprocessing.vectorized_base_image_augmentation_layer import (
+    VectorizedBaseImageAugmentationLayer,
 )
 
 
 @tf.keras.utils.register_keras_serializable(package="keras_cv")
-class Grayscale(BaseImageAugmentationLayer):
+class Grayscale(VectorizedBaseImageAugmentationLayer):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/grayscale.py:23]: I suppose it's good that this layer's tests don't need to be changed.

Can you please verify the performance of this new solution by A/B testing the old+new grayscale layer?

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Tagging related issues:
#581
#241

File: keras_cv/layers/preprocessing/grayscale.py (hunk 8 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/grayscale.py:23]: Done

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:171]: Oops - updated

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:173]: Typo on my part!  and arg order wrong.

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:153]: Hmm, yeah but thats a follow up PR IMO.  we should remove it from all layers.  francois prefers "targets" because OHE labels are not labels anymore but regression targets - makes a good point but at this point no strong preference.  

In any case follow up.

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:188]: Done

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer_test.py]: Done -- can you paste me the command to accomplish this?

File: benchmarks/vectorized_grayscale.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
+# Copyright 2022 The KerasCV Authors
```
Comment (ianstenbit) [inline benchmarks/vectorized_grayscale.py:1]: 2023

**Review by ianstenbit — APPROVED** [review]
LGTM once the GPU tests are passing.

File: keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+      more keys in future when we support more types of augmentation.
+
+    The output of the `call()` will be in two formats, which will be the same
+    structure as the inputs.
+
+    The `call()` will handle the logic detecting the training/inference mode,
+    unpack the inputs, forward to the correct function, and pack the output back
+    to the same structure as the inputs.
+
+    Note that since the randomness is also a common functionnality, this layer
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/vectorized_base_image_augmentation_layer.py:75]: nit: typo (functionality)

**Comment by LukeWood** [comment]
Grayscale test passed --- root cause is that on GPU tf.image.grayscale is not memory efficient.  in practice this barely matters because in tf.data we force CPU scope.  If users request we can come up with a fix using pure tf-ops instead of tf.image ops
